source/contributed/typescript-and-screeps-2.md

```diff
@@ -0,0 +1,598 @@
+- The `miner` class stores the `id` of the Source it's mining.
+- The `upgrader` class stores the `pos` of the controller in memory.
+
+These files are organized in the following directory structure:
+```
+src/
+  workers/
+    worker.ts
+    miner.ts
+    upgrader.ts
+```
+An example of how this code might look is:
+
+```typescript
+// src/workers/worker.ts
+export abstract class Worker {
+  public creep: Creep;
+  public memory: any;
+  
+  constructor(creep: Creep) {
+    this.creep = creep;
+    this.memory = Memory.creeps[creep.name];
+  }
+  
+  // the following would actually be better implemented
+  // as a getter/setter, or as a function on inherited classes
+  // but this is only an example
+  public getJob() {
+    return this.memory.job;
+  }
+  
+  public setJob(job: string) {
+    this.memory.job = job;
+  }
+}
+
+// src/workers/miner.ts
+export class Miner extends Worker {
+
+  constructor(creep: Creep) {
+    super(creep);
+    this.memory.job = "Miner";
+  }
+
+  public setTargetSource(source: Source) {
+    this.memory.source = source.id;
+  }
+  
+  public harvest() {
+    this.creep.harvest(this.memory.source);
+    // Uh Oh! This compiled fine, but will error at runtime!
+    // `this.memory.source` is an ID (string), not an actual Source object
+    // ... If only TypeScript could have warned us....
+    // (not to mention, it could also be `undefined`)
+  }
+}
+```
+
+As our code gets more advanced, we create more classes that inherit from `Worker`, and add more features to the base, we start to lose concept of what exactly is in Memory.
+
+The TS compiler's main advantage is situations like this.  We waste the #1 advantage of using TypeScript when we type objects as `any`.
+
+> **Note** There is [PR](https://github.com/screepers/Screeps-Typescript-Declarations/pull/107) in the [Screeps-Typescript-Declarations](https://github.com/screepers/Screeps-Typescript-Declarations) to fix this.  Until the PR is merged, you can either manually edit the installed typings as in the guide here, or reinstall the typings from the [forked repository](https://github.com/bryanbecker/Screeps-Typescript-Declarations).
+
+A better declaration for Memory would be:
+
+```typescript
+// src/typings/globals/screeps/index.d.ts
+interface Memory {
+    [name: string]: any;
+    creeps: {[name: string]: CreepMemory};
+    flags: {[name: string]: FlagMemory};
+    rooms: {[name: string]: RoomMemory};
+    spawns: {[name: string]: SpawnMemory};
+}
+
+// individual memory interfaces that we will extend later in our codebase
+interface CreepMemory {}
+interface FlagMemory {}
+interface RoomMemory {}
+interface SpawnMemory {}
+```
+
+Now we can start actually telling TypeScript about all the information we are storing in our Creeps Memory.
+
+> **Note** tslint will complain about having an empty interface here, as usually that is a symptom of some other problem.  However, in our case it's nice to help keep track of which object we are working with.
```

Comment:
This depends entirely on your setup, some might not use tslint or simply run with a different configuration.